### PR TITLE
feat(stream): support tombstone row as backfill position

### PR DESCRIPTION
**Pull request by kwannoel**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

https://github.com/risingwavelabs/risingwave/issues/12680

Currently when backfill attempts to `store_create_iter` for its `snapshot_read`, it could stall, if the leading rows are tombstones.

This PR makes the backfill aware of tombstone rows, so it can use them when computing its current_position.

Backfill will now receive tombstone rows for state store's iterator. Then, it will maintain a separate single-row buffer for the latest tombstone row.
At the end of the epoch, it will compare this row to the latest row from non-tombstone rows. If this row has a higher pk, it will use it as its current_pos.

We separate it from the collected chunks, since tombstone rows shouldn't be yielded downstream. Setting visibility to invisible for tombstone rows is also an option, but it makes the cardinality of the chunk large, and could lead to OOM.

In all, now tombstone rows can also occupy the current_pos, but won't be yielded downstream.

This makes sure that in the scenario that the current epoch has long runs of tombstone rows, in the next epoch, we can resume progress from the last tombstone row, instead of starting from scratch, as is currently the case.

In terms of correctness, since tombstone rows are treated as current_pos, at the end of the epoch, updates which are BEFORE the current_pos will still get flushed downstream.
As such, although these updates are not included in the next snapshot_read (since it will only read records AFTER current_pos), they will be included, since the previous epoch still flushes them downstream.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/storage/src/table/batch_table/storage_table.rs [hunk 1]
```diff
@@ -631,8 +635,36 @@ impl<S: StateStore, SD: ValueRowSerde> StorageTableInner<S, SD> {
...
+            ordered,
+            prefetch_options,
+            false,
+        )
+        .await
+    }
+
+    /// Construct a [`StorageTableInnerIter`] for batch executors.
+    /// Differs from the streaming one, this iterator will wait for the epoch before iteration
+    pub async fn iter_tombstone_with_pk_bounds(
```
Comment (kwannoel) [inline src/storage/src/table/batch_table/storage_table.rs:651]: Relational interface.

File: src/storage/src/table/batch_table/storage_table.rs [hunk 2]
```diff
@@ -461,6 +462,7 @@ impl<S: StateStore, SD: ValueRowSerde> StorageTableInner<S, SD> {
                     read_version_from_backup: read_backup,
                     prefetch_options,
                     cache_policy,
+                    with_tombstone,
```
Comment (kwannoel) [inline src/storage/src/table/batch_table/storage_table.rs:465]: Pass in `with_tombstone` to state_store layer via `ReadOptions`.

**Comment by Little-Wallace** [comment]
Please add more description for this PR. 
Because there may be range tombstone in our system. And it is important to tell user how range tombstone will  be in order of output keys flow.
I want to know why we need to support such a interface because tombstone keys flow is not stable since every compaction job would change the result.

**Comment by kwannoel** [comment]
> Please add more description for this PR. Because there may be range tombstone in our system. And it is important to tell user how range tombstone will be in order of output keys flow. I want to know why we need to support such a interface because tombstone keys flow is not stable since every compaction job would change the result.

Hi, I have provided more context in the PR description.

File: src/storage/src/table/mod.rs [hunk 3]
```diff
@@ -148,6 +148,33 @@ where
...
+where
+    S: Stream<Item = Result<KeyedRow<Bytes>, E>> + Unpin,
+{
+    // TODO(kwannoel): If necessary, we can optimize it in the future.
+    // This can be done by moving the check if builder is full from `append_one_row` to here,
+    while let Some(row) = stream.next().await.transpose()? {
+        if row.is_tombstone {
+            *tombstone_row = Some(row.into_owned_row());
+            continue;
+        }
```
Comment (kwannoel) [inline src/storage/src/table/mod.rs:166]: Just need to maintain a single tombstone row, since the stream must be sorted.

**Comment by kwannoel** [comment]
Encounter this error in my testing:
```
thread 'risingwave-streaming-actor' panicked at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/bytes-1.5.0/src/buf/buf_impl.rs:289:9:
assertion failed: self.remaining() >= 1
stack backtrace:
   0: rust_begin_unwind
             at /rustc/62ebe3a2b177d50ec664798d731b8a8d1a9120d1/library/std/src/panicking.rs:619:5
   1: core::panicking::panic_fmt
             at /rustc/62ebe3a2b177d50ec664798d731b8a8d1a9120d1/library/core/src/panicking.rs:72:14
   2: core::panicking::panic
             at /rustc/62ebe3a2b177d50ec664798d731b8a8d1a9120d1/library/core/src/panicking.rs:127:5
   3: bytes::buf::buf_impl::Buf::get_u8
             at /Users/noelkwan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/bytes-1.5.0/src/buf/buf_impl.rs:289:9
   4: <risingwave_common::util::value_encoding::column_aware_row_encoding::Deserializer as risingwave_common::util::value_encoding::ValueRowDeserializer>::deserialize
             at ./src/common/src/util/value_encoding/column_aware_row_encoding.rs:187:36
   5: <risingwave_common::util::value_encoding::column_aware_row_encoding::ColumnAwareSerde as risingwave_common::util::value_encoding::ValueRowDeserializer>::deserialize
             at ./src/common/src/util/value_encoding/column_aware_row_encoding.rs:268:9
   6: <risingwave_common::util::value_encoding::EitherSerde as risingwave_common::util::value_encoding::ValueRowDeserializer>::deserialize
             at ./src/common/src/util/value_encoding/mod.rs:77:33
   7: risingwave_storage::table::batch_table::storage_table::StorageTableInnerIterInner<S,SD>::into_stream::{{closure}}
             at ./src/storage/src/table/batch_table/storage_table.rs:761:28
```

**Comment by hzxa21** [comment]
> storage_table

I pushed a fix to resolve this panic.

**Comment by kwannoel** [comment]
Resolved by https://github.com/risingwavelabs/risingwave/pull/12780

**Pull request closed without merging**